### ISSUE TITLE
A .well-known client config containing a `m.homeserver` field is unnecessarily required in order to use a homeserver

### ISSUE THREAD

**Issue by anoadragon453** [vector-im/element-x-android#137]
### Steps to reproduce

1. Put the domain name of a homeserver that does not have a .well-known client config in the server selection box on the login page, such as `ggc-project.de`.
2. ElementX produces an error, saying the homeserver is invalid.

Edit: @manuroe argues below that a well-known config will eventually be temporarily required for ElementX (which will only ever use Sliding Sync) in order for the app to discover a homeserver's sliding sync proxy (until Sliding Sync is standardised, in which case it will exist at `/_matrix/client/v4/sync`). ElementX Android currently hardcodes the Sliding Sync proxy URL however, but one presumes it will eventually discover it from the homeserver's .well-known config (or allow the user to optionally input it in the UI for now?). 

This issue concerns ElementX refusing to acknowledge a homeserver as valid unless it as a `m.homeserver` field in the .well-known client file, which should not be required.

ggc-project.de is a valid homeserver address (see https://ggc-project.de/_matrix/client/versions), but it is not using [delegation](https://spec.matrix.org/v1.6/server-server-api/#server-discovery). Thus, it is not necessary or particularly useful for the homeserver to have a `.well-known` client config, as the homeserver's client-server API is available under https://ggc-project.de already.

#### What did you expect?

Typing `ggc-project.de` into the homeserver selection box should just work. If a well-known client config is not provided by the homeserver, then the app should just attempt to query `/_matrix/client/versions` against the provided URL (so query `ggc-project.de/_matrix/client/versions` and check for a valid response).

#### What happened instead?

ElementX produces an error and claims that the homeserver is invalid. ElementX currently requires a well-known client config with the `m.homeserver` key.

### Your phone model

Google Pixel 4a 5G

### Operating system version

Android 13.0

### Will you send logs?

No

### Are you willing to provide a PR?

No

**Comment by manuroe** [comment]
ElementX apps currently require a well-known file to discover which sliding sync proxy to use. EX apps only work with a SS proxy.

**Comment by MTRNord** [comment]
> ElementX apps currently require a well-known file to discover which sliding sync proxy to use. EX apps only work with a SS proxy.

I wonder if it would make sense in the MSC to "standardize" a location for the proxy as part of the unstable prefix until the whole thing is stabilized since there are scenarios in which a well-known may be very hard or impossible to set up. From what I understand the proxy is going to stay around for a while anyway.

**Comment by anoadragon453** [comment]
@MTRNord Could you leave that as a comment on MSC3575?

**Comment by anoadragon453** [comment]
> ElementX apps currently require a well-known file to discover which sliding sync proxy to use. EX apps only work with a SS proxy.

That makes sense, but currently ElementX will not let you continue if you don't have an `m.homeserver` key in a .well-known client config, which is not related to sliding sync.

**Comment by MTRNord** [comment]
> @MTRNord Could you leave that as a comment on [MSC3575](https://github.com/matrix-org/matrix-spec-proposals/pull/3575)?

Sure :)

**Comment by bmarty** [comment]
In the case of such homeserver, the app is receiving a `RustAuthenticationException.Generic` error from the SDK with the content: `An error occurred: error sending request for url (https://gcc-project.de/.well-known/matrix/client): error trying to connect: dns error: failed to lookup address information: No address associated with hostname`.

We may try to catch this and do further request application side to check if this is a missing `.well-known` file error (by checking for /version for instance), but it's probably better to add a new error SDK side.

**Comment by anoadragon453** [comment]
> No address associated with hostname

That sounds like the website has gone down in the meantime, and is now no longer a suitable test target.

I've modified https://amorgan.xyz/.well-known/matrix/client to temporarily match the test conditions, if you would like to test with that instead.

**Comment by bmarty** [comment]
Same issue:

<img width="304" alt="image" src="https://github.com/vector-im/element-x-android/assets/3940906/cae73008-4193-4c50-a84d-3b2dc9b99676">
<img width="307" alt="image" src="https://github.com/vector-im/element-x-android/assets/3940906/73aabc3e-6ddb-48e5-b98d-737708c26691">

**Comment by anoadragon453** [comment]
Oh sorry, I've gotten this wrong. amorgan.xyz would require a `m.homeserver` field as **it's not a matrix homeserver** (it should point to matrix.amorgan.xyz).

The problem is that **if you put matrix.amorgan.xyz in the field** (which doesn't need a `m.homeserver` field), then you'll find that ElementX requires a `m.homeserver` field, even though it's not necessary.

I've now modified matrix.amorgan.xyz's .well-known file accordingly. Please test with that instead @bmarty: https://matrix.amorgan.xyz/.well-known/matrix/client

**Comment by anoadragon453** [comment]
I've tested manually on the latest commit (https://github.com/vector-im/element-x-android/commit/1c15db26868d64c74f7027cb87321eae4c4bd735) and Element X will not recognise a homeserver as being valid if it has a well-known file but that file does not contain a `m.homeserver` key.

However, I also checked the Client-Server spec for `GET /.well-known/matrix/client` and saw that this field is in fact **required**, so a client not working properly if it isn't there is entirely expected :facepalm: 

Therefore I think Element X is acting correctly in this case. Sorry for the time sink!
